### PR TITLE
search for artifacts that contain a given class

### PR DESCRIPTION
**Pull request by maddingo**
I changed the implementation after your refactoring. I also found it useful to create a Query builder and create the query otside the client

### CODE REVIEW

File: src/test/java/it/mulders/mcs/search/WildcardSearchQueryTest.java [hunk 1]
```diff
@@ -13,30 +13,17 @@
     @DisplayName("withLimit")
     class WithLimitTest {
         @Test
-        void null_limit_should_return_original_object() {
-            var query = new WildcardSearchQuery("foo");
+        void simple_query_should_return_wildcard_query() {
```
Comment (mthmulders) [inline src/test/java/it/mulders/mcs/search/WildcardSearchQueryTest.java:16]: I like this method name: very clear as to what you can expect the test to verify :-)

File: src/main/java/it/mulders/mcs/search/SearchQuery.java [hunk 2]
```diff
@@ -1,36 +1,73 @@
...
-                var version = parts[2];
-                return new CoordinateQuery(groupId, artifactId, version);
-            } else {
-                return new CoordinateQuery(groupId, artifactId);
+                    var groupId = parts[0];
+                    var artifactId = parts[1];
+                    var version = parts.length == 3 ? parts[2] : null;
+                    return new CoordinateQuery(groupId, artifactId, version, limit != null ? limit : Constants.DEFAULT_MAX_SEARCH_RESULTS);
+                } else {
+                    return new WildcardSearchQuery(query, limit != null ? limit : 1);
```
Comment (mthmulders) [inline src/main/java/it/mulders/mcs/search/SearchQuery.java:63]:
```suggestion
                    return new WildcardSearchQuery(query, limit != null ? limit : Constants.DEFAULT_MAX_SEARCH_RESULTS);
```

**Review by mthmulders — CHANGES_REQUESTED** [review]
Thanks - again! - for your contribution and your patience. I do like some of the ideas you've introduced around the `SearchQuery` interface and its implementations, but there are also a few things I'd like to give another thought.

I can't yet say how that will turn out. How about if I have an idea before this PR is merged, we will solve it in this PR - otherwise I'll refactor it myself later?

File: src/main/java/it/mulders/mcs/cli/Cli.java [hunk 3]
```diff
@@ -65,9 +70,42 @@ public SearchCommand() {
...
+                }
+        )
+        private String query;
+
+        @CommandLine.Option(
+                names = { "-f", "--full-name" },
+                negatable = true,
+                arity = "0",
+                description = "Show <count> last versions",
+                paramLabel = "<count>"
```
Comment (mthmulders) [inline src/main/java/it/mulders/mcs/cli/Cli.java:100]: I think it makes sense to have the `"-l" / "--last"` parameter for class-name search, too?

**Comment by mthmulders** [comment]
I gave it a bit of thought. Again, I like the idea of the `SearchQuery` interface and having a builder for it. What I think I'd like to improve is two things

- the builder now knows everything about all types of search queries. I think I'd rather pull that apart into separate builders for each type of query
- the two queries for class-search are almost identical. I think I'd rather merge them.

If you want to, I can create a pull request against your branch, so you can evaluate what you think of these ideas?

**Comment by maddingo** [comment]
Hi Maarten,
you go ahead and pull it apart as much as you like. I wasn't sure how to do
it properly in the spirit of the project. I tried though. I am glad you
like some of the ideas. Just use them at your leisure.
The class search and the full-class search are almost identical, 'c:' vs.
'fc:'. 'fc:' is very useful, e.g. if you go from J2EE to spring-boot and
have to find the artifact that contains a missing class.
Thanks for putting this tool in sdkman.
--
Martin

**Comment by mthmulders** [comment]
> you go ahead and pull it apart as much as you like. I wasn't sure how to do it properly in the spirit of the project. I tried though. I am glad you like some of the ideas. Just use them at your leisure.

I definitely like the idea and look forward to including it in the project. Are you suggesting I take the feature from here and blend it into the code base?

> The class search and the full-class search are almost identical, 'c:' vs. 'fc:'. 'fc:' is very useful, e.g. if you go from J2EE to spring-boot and have to find the artifact that contains a missing class.

That's a good use case, I wouldn't even have thought of that. The same probably applies if you go from `javax.*` to `jakarta.*` namespaces.

> Thanks for putting this tool in sdkman.

You're welcome! Glad you find it useful.

**Comment by maddingo** [comment]
Hi Maarten,
yes, just take what you need and blend it into your code. Don't bother with
attribution or anything. It is all yours. The best is probably to reject
the PR after you have taken what you need. Sorry for not being more
cooperative.

Martin

**Comment by mthmulders** [comment]
> Sorry for not being more cooperative.

Not to worry. I appreciate the idea for this feature, I appreciate the effort you took in figuring out the API for it, and I appreciate the ideas you gave for refactoring!

> Yes, just take what you need and blend it into your code.

I'll take it from here.

**Pull request closed without merging**